pydantic/_internal/_discriminated_union.py

```diff
@@ -448,41 +449,37 @@ def _infer_discriminator_values_for_inner_schema(
         """When inferring discriminator values for a field, we typically extract the expected values from a literal
         schema. This function does that, but also handles nested unions and defaults.
         """
-        if schema['type'] == 'literal':
+        schema_type = schema['type']
+
+        if schema_type == 'literal':
             return schema['expected']
 
-        elif schema['type'] == 'union':
-            # Generally when multiple values are allowed they should be placed in a single `Literal`, but
-            # we add this case to handle the situation where a field is annotated as a `Union` of `Literal`s.
-            # For example, this lets us handle `Union[Literal['key'], Union[Literal['Key'], Literal['KEY']]]`
-            values: list[Any] = []
+        if schema_type == 'union':
+            values = []
             for choice in schema['choices']:
                 choice_schema = choice[0] if isinstance(choice, tuple) else choice
                 choice_values = self._infer_discriminator_values_for_inner_schema(choice_schema, source)
                 values.extend(choice_values)
             return values
 
-        elif schema['type'] == 'default':
-            # This will happen if the field has a default value; we ignore it while extracting the discriminator values
+        if schema_type == 'default':
             return self._infer_discriminator_values_for_inner_schema(schema['schema'], source)
```

Comment:
```suggestion
        if schema_type == 'default':
            # This will happen if the field has a default value; we ignore it while extracting the discriminator values
            return self._infer_discriminator_values_for_inner_schema(schema['schema'], source)
```